R/functions.R

```diff
@@ -0,0 +1,345 @@
+# Functions used to help in the planning googlesheet ----------------------
+
+#' Return id of the planning googlesheet
+gs_id <- function() "1HLtyGK_csi5W_v7XChxgTuVjS-RKXqc0Jxos1RBqpwk"
+
+#' Generate a long-format planning table from the Planning_v2 sheet in the
+#' planning googlesheet.
+#'
+#' Reads the 'wide-format' Planning_v2 sheet from the planning googlesheet and
+#' tries to update the planned month intervals to actual (expected/required)
+#' month intervals, by taking into account current task status, current date,
+#' actual month that the task was started.
+#' The 'long-format' part is about several things: putting persons below each
+#' other, putting 'uitvoering' and 'review' below each other, and also the
+#' expansion of intervals to the respective months.
+#'
+#' @param ss The id of the planning googlesheet
+get_planning_long <- function(ss = gs_id()) {
+  # read the planning data
+  read_sheet(
+    ss,
+    sheet = "Planning_v2",
+    col_types = "cccclllilccccccdddddddddddccdddddddddddcc",
+    .name_repair = "minimal"
+  ) |>
+    # clean planning data and turn it into long format
+    clean_names() |>
+    select(
+      -uitvoerders,
+      -reviewers,
+      -starts_with("totaal_dagen"),
+      -ends_with("_buiten_kern"),
+      -starts_with("tijdsinvestering")
+    ) |>
+    rename_with(
+      .cols = karen:datamanager,
+      .fn = \(x) {
+        str_c(x, "_1")
+      }
+    ) |>
+    mutate(
+      continuous = deadline == "Doorlopend",
+      deadline = ifelse(
+        deadline == "Doorlopend",
+        paste0(max_y(), "-12"),
+        deadline
+      ),
+      # intended dates
+      start = ym(start),
+      deadline = ym(deadline) + months(1) - days(1),
+      # actual months (as dates) of starting and finishing (only for started tasks)
+      gestart = ym(gestart),
+      afgerond = ym(afgerond) + months(1) - days(1),
+      # derived variables:
+      # ####################
+      # 2 logicals
+      started = !is.na(gestart),
+      finished = !is.na(afgerond),
+      # first & last day of current month
+      first_day_currentmonth = floor_date(today(), unit = "month"),
+      last_day_currentmonth = ceiling_date(today(), unit = "month") - days(1),
+      # which start date to take in calculating currently applicable interval?
+      begin = case_when(
+        started ~ gestart,
+        !started & finished ~ start,
+        .default = pmax(start, first_day_currentmonth, na.rm = TRUE)
+      ),
+      # which end date to take in calculating currently applicable interval?
+      end = ifelse(finished, afgerond, deadline) |>
+        as.Date(),
+      # is the task overdue?
+      overdue = !finished & today() > end,
+      # number of months of the interval
+      nr_months = case_when(
+        started & overdue ~ as.period(last_day_currentmonth - begin),
```

Comment:
As it happens, Karen was inserting a half-empty line at the moment of review.
This crashed the `uncount` below since `Start` and `Deadline`, and in consequence `nr_months` were `NA`.
I'd suggest to insert:
```
        is.na(start) & is.na(deadline) ~ period(0,"month"),
```
here.

Reply:
Good idea. Maybe it's even better to trigger an error in such case: such cases should not exist. You're free to implement that.

Reply:
I think the best is what you propose, but accompanied by a warning. So that code still runs in case the googlesheet cannot be fixed immediately.